extensions/jdbc/jdbc-mariadb/runtime/src/main/java/io/quarkus/jdbc/mariadb/runtime/graal/AuthenticationPluginLoader_Substitutions.java

```diff
@@ -2,51 +2,41 @@
 
 import java.sql.SQLException;
 
-import org.mariadb.jdbc.authentication.AuthenticationPlugin;
-import org.mariadb.jdbc.authentication.AuthenticationPluginLoader;
-import org.mariadb.jdbc.internal.com.send.authentication.ClearPasswordPlugin;
-import org.mariadb.jdbc.internal.com.send.authentication.Ed25519PasswordPlugin;
-import org.mariadb.jdbc.internal.com.send.authentication.NativePasswordPlugin;
-import org.mariadb.jdbc.internal.com.send.authentication.OldPasswordPlugin;
-import org.mariadb.jdbc.internal.com.send.authentication.SendGssApiAuthPacket;
+import org.mariadb.jdbc.Configuration;
+import org.mariadb.jdbc.plugin.AuthenticationPlugin;
+import org.mariadb.jdbc.plugin.authentication.addon.ClearPasswordPlugin;
+import org.mariadb.jdbc.plugin.authentication.addon.SendGssApiAuthPacket;
+import org.mariadb.jdbc.plugin.authentication.standard.CachingSha2PasswordPlugin;
+import org.mariadb.jdbc.plugin.authentication.standard.Ed25519PasswordPlugin;
+import org.mariadb.jdbc.plugin.authentication.standard.NativePasswordPlugin;
 
 import com.oracle.svm.core.annotate.Substitute;
 import com.oracle.svm.core.annotate.TargetClass;
 
-@TargetClass(AuthenticationPluginLoader.class)
-@Substitute
+@TargetClass(org.mariadb.jdbc.plugin.authentication.AuthenticationPluginLoader.class)
 public final class AuthenticationPluginLoader_Substitutions {
 
-    public static final String MYSQL_NATIVE_PASSWORD = "mysql_native_password";
-    public static final String MYSQL_OLD_PASSWORD = "mysql_old_password";
-    public static final String MYSQL_CLEAR_PASSWORD = "mysql_clear_password";
-    public static final String MYSQL_ED25519_PASSWORD = "client_ed25519";
-    private static final String GSSAPI_CLIENT = "auth_gssapi_client";
-    private static final String DIALOG = "dialog";
-
     @Substitute
-    public static AuthenticationPlugin get(String type) throws SQLException {
+    public static AuthenticationPlugin get(String type, Configuration conf) throws SQLException {
         switch (type) {
-            case MYSQL_NATIVE_PASSWORD:
+            case NativePasswordPlugin.TYPE:
                 return new NativePasswordPlugin();
-            case MYSQL_OLD_PASSWORD:
-                return new OldPasswordPlugin();
-            case MYSQL_CLEAR_PASSWORD:
+            case ClearPasswordPlugin.TYPE:
                 return new ClearPasswordPlugin();
-            case DIALOG:
+            case "dialog":
                 throw new UnsupportedOperationException("Authentication strategy 'dialog' is not supported in GraalVM");
             //return new SendPamAuthPacket();
-            case GSSAPI_CLIENT:
+            case "auth_gssapi_client":
                 return new SendGssApiAuthPacket();
-            case MYSQL_ED25519_PASSWORD:
+            case "client_ed25519":
                 return new Ed25519PasswordPlugin();
-
+            case CachingSha2PasswordPlugin.TYPE:
+                return new CachingSha2PasswordPlugin();
```

Comment:
This one is new in 3.0 and it didn't look "native-hostile" to me but I didn't test it!

Reply:
Perhaps this whole method needs to be updated to align with the latest version? In particular our code doesn't implement the check for the selected plugin to be compatible with the configured restrictions listed in `conf.restrictedAuth()`, that would make me a bit nervous as it opens a possible security weakness.

Reply:
Good point! `restrictedAuth()` is new in 3.0 and we should definitely respect it.

I'm not comfortable though with c&p of the entire logic. I'll try to have the plugins loaded during build time so that we don't need to substitute that method at all.
If you have hints on that please let me know.

Reply:
To me this looks like the best solution;
```java
    @BuildStep
    void registerAuthenticationPlugins(BuildProducer<ServiceProviderBuildItem> serviceProvider) {
        // make sure that all plugins are available
        serviceProvider
                .produce(ServiceProviderBuildItem.allProvidersFromClassPath("org.mariadb.jdbc.plugin.AuthenticationPlugin"));
    }
```
plus _not_ `@Delete`ing `SendPamAuthPacket` entirely but substituting its method bodies with `UnsupportedOperationException`.

This way we can leave `AuthenticationPluginLoader` untouched.

Reply:
> This way we can leave `AuthenticationPluginLoader` untouched.

I pushed that solution.